src/main/java/spoon/support/compiler/jdt/ContextBuilder.java

```diff
@@ -112,55 +125,153 @@ void exit(ASTNode node) {
 		}
 	}
 
+	@SuppressWarnings("unchecked")
 	<T> CtLocalVariable<T> getLocalVariableDeclaration(final String name) {
-		for (ASTPair astPair : this.stack) {
-			// TODO check if the variable is visible from here
+		final Class<CtLocalVariable<T>> clazz = (Class<CtLocalVariable<T>>)
+				jdtTreeBuilder.getFactory().Core().createLocalVariable().getClass();
+		final CtLocalVariable<T> localVariable =
+				this.<T, CtLocalVariable<T>>getVariableDeclaration(name, clazz);
+		if (localVariable == null) {
+			// note: this happens when using the new try(vardelc) structure
+			this.jdtTreeBuilder.getLogger().error(
+					format("Could not find declaration for local variable %s at %s",
+							name, stack.peek().element.getPosition()));
+		}
+		return localVariable;
+	}
+
+	@SuppressWarnings("unchecked")
+	<T> CtCatchVariable<T> getCatchVariableDeclaration(final String name) {
+		final Class<CtCatchVariable<T>> clazz = (Class<CtCatchVariable<T>>)
+				jdtTreeBuilder.getFactory().Core().createCatchVariable().getClass();
+		final CtCatchVariable<T> catchVariable =
+				this.<T, CtCatchVariable<T>>getVariableDeclaration(name, clazz);
+		if (catchVariable == null) {
+			// note: this happens when using the new try(vardelc) structure
+			this.jdtTreeBuilder.getLogger().error(
+					format("Could not find declaration for catch variable %s at %s",
+							name, stack.peek().element.getPosition()));
+		}
+		return catchVariable;
+	}
 
-			EarlyTerminatingScanner<CtLocalVariable<?>> scanner = new EarlyTerminatingScanner<CtLocalVariable<?>>() {
+	<T> CtVariable<T> getVariableDeclaration(final String name) {
+		final CtVariable<T> variable = this.<T, CtVariable<T>>getVariableDeclaration(name, null);
+		if (variable == null) {
+			// note: this happens when using the new try(vardelc) structure
+			this.jdtTreeBuilder.getLogger().error(
+					format("Could not find declaration for variable %s at %s",
+							name, stack.peek().element.getPosition()));
+		}
+		return variable;
+	}
+
+	@SuppressWarnings("unchecked")
+	private <T, U extends CtVariable<T>> U getVariableDeclaration(
+			final String name, final Class<U> clazz) {
+		final CoreFactory coreFactory = jdtTreeBuilder.getFactory().Core();
+		final ClassFactory classFactory = jdtTreeBuilder.getFactory().Class();
+		final InterfaceFactory interfaceFactory = jdtTreeBuilder.getFactory().Interface();
+		final ReferenceBuilder referenceBuilder = jdtTreeBuilder.getReferencesBuilder();
+		// there is some extra work to do if we are looking for CtFields (and subclasses)
+		final boolean lookingForFields = clazz == null
+				|| coreFactory.createField().getClass().isAssignableFrom(clazz);
+
+		// try to find the variable on stack beginning with the most recent element
+		for (final ASTPair astPair : stack) {
+			// the variable may have been declared directly by one of these elements
+			final List<U> variables = astPair.element.getElements(new AbstractFilter<U>() {
```

Comment:
I think there was a reason but let me have a look if an EarlyTerminatingScanner is feasible.